In [1]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - Latest 12 weeks of life of the banana plant
---

In [2]:
temperature_l12w = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/Latest_12-weeks_May-4_August-03_2018/' \
                                 'Temperatura_May-4_August-03_2018.csv', )

In [3]:
print(temperature_l12w.shape)
temperature_l12w.head()

(4398, 2)


Fecha:  Temperatura (C)
0  2018-05-04 00:07:38            23.75
1  2018-05-04 00:37:39            23.68
2  2018-05-04 01:07:38            23.68
3  2018-05-04 01:37:39            23.56
4  2018-05-04 02:07:39            23.50

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [4]:
print(temperature_l12w.isnull().any())
temperature_l12w.isnull().values.any()

Fecha:             False
Temperatura (C)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Temperatura (C)` ** and assigning them to the matrix 

`temperature_l12w_array` created such as follow:

In [5]:
temperature_l12w_array = temperature_l12w.iloc[:,1].values

In [6]:
# Rehape the luminosity_luxes array 
temperature_l12w_array = temperature_l12w_array.reshape(-1, 1)
temperature_l12w_array

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

## 1.3 Generating descriptive data to dataset
---

In [7]:
col = ['Temperatura (C)']
temperature_l12w_values_df = pd.DataFrame(temperature_l12w_array, columns=col)
temperature_l12w_values_df_describe = temperature_l12w_values_df.describe()

In [8]:
temperature_l12w_values_df_describe

Temperatura (C)
count      4398.000000
mean         27.531389
std           4.273432
min          21.370000
25%          24.000000
50%          25.900000
75%          31.060000
max          40.250000

In [9]:
# Export this descriptive data to comma separated values and java script object notation
temperature_l12w_values_df_describe.to_csv('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                              'Temperature_Describe_May-4_August-03_2018.cvs', sep=',', header=True, index=True)
temperature_l12w_values_df_describe.to_json('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n'
                                     'Temperature_Describe_May-4_August-03_2018.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`temperature_l12w_values_df`** dataset with 4398 samples rows.

In [10]:
print(temperature_l12w_values_df.shape)
temperature_l12w_values_df.head()

(4398, 1)


Temperatura (C)
0            23.75
1            23.68
2            23.68
3            23.56
4            23.50

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`temperature_l12w_values_df`** dataframe to numpy array such as follow:

In [11]:
# numpy_temperature_values = temperature_values_df.reset_index().values
numpy_temperature_l12w_values = temperature_l12w_values_df.values

In [12]:
numpy_temperature_l12w_values

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

We compose the following datasets from **`numpy_temperature_l12w_values`** array :

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_temperature_l12w_values`** which contain all column features of temperature.

Of this way, we pass all data (4390 samples rows) to from them `temperature_l12w_values_train`

and `temperature_l12w_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **4390 * 0.2 = 878 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **4390 * 0.8 = 3512  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing temperature datasets
---

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

In [13]:
temperature_l12w_values_train, temperature_l12w_values_test = train_test_split(numpy_temperature_l12w_values, test_size = 0.2)

In [14]:
# We have 3518 rows to luminosity_luxes_train
print(type(temperature_l12w_values_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , temperature_l12w_values_train.shape)
print('\n')

# And we have 880 rows to luminosity_luxes_test
print(type(temperature_l12w_values_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , temperature_l12w_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (3518, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (880, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations


---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---

We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Temperature (C)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Temperature training dataset

In [15]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', temperature_l12w_values_train)

Remember our temperature training data 
 [[22.25]
 [27.62]
 [33.5 ]
 ...
 [23.81]
 [24.56]
 [25.06]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `temperature_l12w_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [16]:
minmax_scale_training = scaler.fit(temperature_l12w_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `temperature_l12w_values_train` data are scaled between **0 to 1**  selected range 

In [17]:
# transform precipitations_train data to maximum and mínimum scale value. 
temperature_l12w_minmax_training = minmax_scale_training.transform(temperature_l12w_values_train)

In [18]:
print ("And, these are our scaled data: " + '\n')
temperature_l12w_minmax_training

And, these are our scaled data: 



array([[0.04661017],
       [0.33103814],
       [0.64247881],
       ...,
       [0.12923729],
       [0.16896186],
       [0.19544492]])

In [19]:
print('Temperature Training dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].min()))

print('Temperature Training dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].max()))

Temperature Training dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Training dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_training` numpy array

- We export this array to comma separated values 


In [20]:
temperature_l12w_train_df = pd.DataFrame(temperature_l12w_minmax_training, columns=col)

In [21]:
temperature_l12w_train_df

Temperatura (C)
0            0.046610
1            0.331038
2            0.642479
3            0.129237
4            0.589513
5            0.046610
6            0.102754
7            0.082627
8            0.188559
9            0.132415
10           0.268008
11           0.592691
12           0.675318
13           0.221928
14           0.658898
15           0.638771
16           0.274894
17           0.731462
18           0.042903
19           0.605932
20           0.539725
21           0.095869
22           0.784428
23           0.063030
24           0.668962
25           0.284958
26           0.347458
27           0.781250
28           0.728284
29           0.496822
...               ...
3488         0.609110
3489         0.294492
3490         0.778072
3491         0.629237
3492         0.549788
3493         0.079449
3494         0.195445
3495         0.172140
3496         0.198623
3497         0.721928
3498         0.284958
3499         0.158898
3500         0.258475
3501         0.281250
3502         0.506356
3503         0.119174
3504         0.506356
3505         0.344280
3506         0.486758
3507         0.099576
3508         0.099576
3509         0.152542
3510         0.092691
3511         0.079449
3512         0.126059
3513         0.132415
3514         0.175318
3515         0.129237
3516         0.168962
3517         0.195445

[3518 rows x 1 columns]

In this way we have the dataset `temperature_l12w_train_df` standardized and training, and export it to a .csv file

In [22]:
temperature_l12w_train_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TRAINING_May-4_August-03_2018.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [23]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', temperature_l12w_values_test)

Remember our temperature testing data 
 [[22.62]
 [26.12]
 [24.87]
 [34.06]
 [37.75]
 [24.56]
 [25.43]
 [30.68]
 [34.31]
 [29.12]
 [23.43]
 [24.56]
 [25.  ]
 [32.  ]
 [33.25]
 [30.06]
 [26.12]
 [36.43]
 [24.12]
 [34.62]
 [23.68]
 [23.75]
 [24.12]
 [36.56]
 [24.  ]
 [23.87]
 [29.56]
 [25.75]
 [35.12]
 [24.31]
 [33.18]
 [32.06]
 [35.37]
 [32.56]
 [23.37]
 [26.12]
 [31.75]
 [26.37]
 [28.62]
 [23.87]
 [24.  ]
 [32.43]
 [25.12]
 [30.25]
 [25.87]
 [23.68]
 [24.12]
 [22.81]
 [26.5 ]
 [24.31]
 [24.06]
 [33.93]
 [23.68]
 [24.5 ]
 [26.06]
 [26.43]
 [23.81]
 [24.06]
 [29.37]
 [28.93]
 [33.43]
 [25.75]
 [24.75]
 [27.81]
 [24.62]
 [37.06]
 [25.31]
 [24.06]
 [24.25]
 [23.  ]
 [35.56]
 [22.81]
 [29.37]
 [23.87]
 [36.56]
 [23.81]
 [29.75]
 [33.68]
 [22.93]
 [23.18]
 [32.37]
 [34.31]
 [25.56]
 [23.43]
 [25.56]
 [32.31]
 [27.43]
 [26.  ]
 [26.68]
 [32.81]
 [26.68]
 [31.18]
 [25.68]
 [25.87]
 [23.68]
 [29.  ]
 [30.81]
 [26.  ]
 [27.25]
 [27.37]
 [24.06]
 [24.62]
 [36.87]
 [34.68]
 [27.43]
 [32.56]
 [34.5

In [24]:
minmax_scale_test = scaler.fit(temperature_l12w_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
temperature_l12w_minmax_test = minmax_scale_test.transform(temperature_l12w_values_test)

In [25]:
print ("And, these are our testing scaled data: " + '\n')
temperature_l12w_minmax_test

And, these are our testing scaled data: 



array([[0.04548007],
       [0.24199888],
       [0.17181359],
       [0.68781583],
       [0.89500281],
       [0.15440764],
       [0.2032566 ],
       [0.49803481],
       [0.70185289],
       [0.41044357],
       [0.09096013],
       [0.15440764],
       [0.17911286],
       [0.57215048],
       [0.64233577],
       [0.46322291],
       [0.24199888],
       [0.82088714],
       [0.12970241],
       [0.71925884],
       [0.10499719],
       [0.10892757],
       [0.12970241],
       [0.82818641],
       [0.12296463],
       [0.11566536],
       [0.43514879],
       [0.22122403],
       [0.74733296],
       [0.14037058],
       [0.63840539],
       [0.57551937],
       [0.76137002],
       [0.60359349],
       [0.08759124],
       [0.24199888],
       [0.55811342],
       [0.25603593],
       [0.38236946],
       [0.11566536],
       [0.12296463],
       [0.59629422],
       [0.18585065],
       [0.47389107],
       [0.22796182],
       [0.10499719],
       [0.12970241],
       [0.056

In [26]:
print('Temperature Testing dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].min()))

print('Temperature Testing dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].max()))

Temperature Testing dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Testing dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_test` numpy array

- We export this array to comma separated values

In [27]:
temperature_l12w_test_df = pd.DataFrame(temperature_l12w_minmax_test, columns=col)

In [28]:
print(temperature_l12w_test_df.shape)
temperature_l12w_test_df

(880, 1)


Temperatura (C)
0           0.045480
1           0.241999
2           0.171814
3           0.687816
4           0.895003
5           0.154408
6           0.203257
7           0.498035
8           0.701853
9           0.410444
10          0.090960
11          0.154408
12          0.179113
13          0.572150
14          0.642336
15          0.463223
16          0.241999
17          0.820887
18          0.129702
19          0.719259
20          0.104997
21          0.108928
22          0.129702
23          0.828186
24          0.122965
25          0.115665
26          0.435149
27          0.221224
28          0.747333
29          0.140371
..               ...
850         0.312184
851         0.122965
852         0.273442
853         0.505334
854         0.477260
855         0.417743
856         0.733296
857         0.575519
858         0.182482
859         0.126334
860         0.445817
861         0.161145
862         0.249298
863         0.126334
864         0.284110
865         0.912409
866         0.045480
867         0.617631
868         0.129702
869         0.501965
870         0.196519
871         0.273442
872         0.754632
873         0.185851
874         0.659742
875         0.151039
876         0.048849
877         0.221224
878         0.256036
879         0.501965

[880 rows x 1 columns]

In [29]:
temperature_l12w_test_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TESTING_May-4_August-03_2018.csv', sep=',', header=True, index=False)